lib/uklibparam/parser.c

```diff
@@ -0,0 +1,707 @@
+#include <uk/version.h>
+#include <uk/essentials.h>
+
+static UK_LIST_HEAD(ld_head);
+
+#define PARSE_PARAM_SEP    '.'
+#define PARSE_VALUE_SEP    '='
+#define PARSE_LIST_START   '['
+#define PARSE_LIST_END     ']'
+#define PARSE_STOP         "--"
+#define PARSE_USAGE        "help"
+
+void _uk_libparam_libsec_register(struct uk_libparam_libdesc *ld)
+{
+	uk_list_add_tail(&ld->next, &ld_head);
+}
+
+static const char *str_param_type(enum uk_libparam_param_type pt)
+{
+	switch (pt) {
+	case _UK_LIBPARAM_PARAM_TYPE_bool:
+		return "bool";
+	case _UK_LIBPARAM_PARAM_TYPE___s8:
+		return "s8";
+	case _UK_LIBPARAM_PARAM_TYPE___u8:
+		return "u8";
+	case _UK_LIBPARAM_PARAM_TYPE___s16:
+		return "s16";
+	case _UK_LIBPARAM_PARAM_TYPE___u16:
+		return "u16";
+	case _UK_LIBPARAM_PARAM_TYPE___s32:
+		return "s32";
+	case _UK_LIBPARAM_PARAM_TYPE___u32:
+		return "u32";
+	case _UK_LIBPARAM_PARAM_TYPE___s64:
+		return "s64";
+	case _UK_LIBPARAM_PARAM_TYPE___u64:
+		return "u64";
+	case _UK_LIBPARAM_PARAM_TYPE___uptr:
+		return "uptr";
+	case _UK_LIBPARAM_PARAM_TYPE_charp:
+		return "charp";
+	default:
+		break;
+	}
+	return "<?>";
+}
+
+#define UK_LIBPARAM_FOREACH_LIBDESC(ld_iter) \
+	uk_list_for_each_entry((ld_iter), &ld_head, next)
+#define UK_LIBPARAM_FOREACH_PARAMIDX(ld, p_iter) \
+	for ((p_iter) = 0; (p_iter) < (ld)->params_len; (p_iter)++)
+#define UK_LIBPARAM_PARAM_GET(libdesc, idx) \
+	(((idx) >= (libdesc)->params_len) ? __NULL : ((libdesc)->params)[idx])
+
+static void uk_usage(void)
+{
+	struct uk_libparam_libdesc *ld;
+	struct uk_libparam_param *p;
+	__sz p_i;
+
+	/*
+	 * FIXME: Use a console print variant without context prefix
+	 *        (Update as soon as it is available)
+	 */
+	uk_pr_info("Usage of command line:\n"
+		   "   [%s] [PREFIX%cPARAMETER%cVALUE]... %s [APPLICATION ARGUMENT]...\n\n",
+		   PARSE_USAGE, PARSE_PARAM_SEP, PARSE_VALUE_SEP, PARSE_STOP);
+	uk_pr_info("Special commands:\n"
+		   "%12s                   Print this help summary\n\n",
+		   PARSE_USAGE);
+	uk_pr_info("Available parameters:\n");
+	UK_LIBPARAM_FOREACH_LIBDESC(ld) {
+		UK_LIBPARAM_FOREACH_PARAMIDX(ld, p_i) {
+			p = UK_LIBPARAM_PARAM_GET(ld, p_i);
+			UK_ASSERT(p);
+
+			if (p->count == 0)
+				continue;
+			uk_pr_info("%12s.%-18s%s (",
+				   ld->prefix, p->name,
+				   p->desc ? p->desc : p->name);
+			if (p->count > 1)
+				uk_pr_info("array[%"__PRIsz"] of ",
+				p->count);
+			uk_pr_info("%s)\n", str_param_type(p->type));
+		}
+	}
+	uk_pr_info("\n"
+		   "Numbers can be passed in decimal, octal (\"0\" as prefix), or hexadecimal (\"0x\" as prefix).\n");
+	uk_pr_info("Valid boolean values for 'true' are: \"true\", \"on\", \"yes\", a non-zero number.\n"
+		   "Valid boolean values for 'false' are: \"false\", \"off\", \"no\", a zero number (e.g., \"0\").\n");
+	uk_pr_info("Boolean parameters that are passed without a value will be set to 'true'.\n");
+	uk_pr_info("Array parameters can be passed with multiple 'LIBRARY%cPARAMETER%cVALUE' tokens,\n",
```

Comment:
Does this `LIBRARY` keyword have another meaning compared to `PREFIX` used the same way (`PREFIX%cPARAMETER%cVALUE`) throughout the help message?
In other words, should I think that I _have_ to use the library name here?

Reply:
Uups, sorry, this was a mistake. It should say PREFIX because this can be defined per-library. In case, no prefix was set up by a library, the prefix becomes the library name.